[Video] Various fixes for bluray:// (and archive://) paths.

78andyp opened the pull request:
## Description

Following on from discussion in #24720 (as amended by discussion below).

Fixes:
- bluray:// path not updated if more than one videoversion
- videoversions would be readded after library scan after bluray:// path assigned
- library clean coud remove bluray:// paths (and archive:// paths)
- now removes orphan `file` and `path` entries
- information diaglog refresh used wrong movie name
- choose playlist option missing if stopped in menu
- import/export of bluray:// containing items
- fix resume when played through menu
- fix could not delete bluray:// when removing from library and rename/delete enabled

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

78andyp commented:
@CrystalP - grateful for your thoughts, thanks.

78andyp commented:
I've played around with versions and extras and can't find a difference between v21.

I do note a couple of ? bugs that are also present in v21
- the artwork for one of the movies is blank in the manage versions dialog (? the selected or first one - I haven't explored further) - it's OK in the select version to play dialog
- when you add a movie that's already been (part) played as a version the movie is deleted and so are the streamdetails (but not a bookmark) - so the stream details aren't shown but you are prompted to resume when you play
- if you add a movie with extras as a version and then delete the movie, the extras specific to the added and then deleted version are still listed when the extras button is pressed.

CrystalP commented:
Thanks for working on a followup.
It will take me a bit of time to do a full review of previous PR + this, but I already have a few comments, not related to the versions yet.
- in CSaveFileState::DoWork() I don't think it's a good idea to systematically call SetFileForMovie/SetFileForEpisode. That should only be done when the idFile changes.
- I don't think that SetStreamDetailsForFile() should be bent into a function returning the idFile of the path. That's not its purpose. Please do the `const int idFile = AddFile(path)` in CSaveFileState::DoWork and then call SetStreamDetailsForFileId() to avoid a double-lookup of the id.
- updating c22 was probably correct, but a search of the rest of the code to check how it's usually populated and used would be needed. There may be some special cases and maybe it's not just a denormalization of the path table.

78andyp commented:
> Thanks for working on a followup. It will take me a bit of time to do a full review of previous PR + this, but I already have a few comments, not related to the versions yet.
> 
> * in CSaveFileState::DoWork() I don't think it's a good idea to systematically call SetFileForMovie/SetFileForEpisode. That should only be done when the idFile changes.
> * I don't think that SetStreamDetailsForFile() should be bent into a function returning the idFile of the path. That's not its purpose. Please do the `const int idFile = AddFile(path)` in CSaveFileState::DoWork and then call SetStreamDetailsForFileId() to avoid a double-lookup of the id.
> * updating c22 was probably correct, but a search of the rest of the code to check how it's usually populated and used would be needed. There may be some special cases and maybe it's not just a denormalization of the path table.

OK, fixed the first two.
It looks like c22 should be left alone - seems to be used to get base path for subtitle and external audio scan, if I'm following the code right.

CrystalP commented:
You may have pushed the wrong commits, I don't see how the first two points are addressed.

78andyp commented:
> You may have pushed the wrong commits, I don't see how the first two points are addressed.

Sorry. I fixed it and then tested it and it didn't work.

Regarding the first point.

The new fileId can get generated in multiple places - (eg.update last played, streamdetails etc..) so it's not really possible to bring it into SaveFileStateJob.

The FileId routines are not being called every time, only if the streamdetails have changed.

I have addressed the second point - I've reverted SetStreamDetails back to a void function. That's how it is in v21.

78andyp commented:
@CrystalP, was this OK??

78andyp commented:
@CrystalP Hi. Was this OK?? I've been using it for a few months now with no ill effects. Thanks.

78andyp commented:
Either should be OK. The end result should be that the chosen playlist (and corresponding streamdetails) are saved, so when played again it's played directly (more important for episodes, esp when playlist numbering is not clear - but there's a seperate PR that makes episode identification more intelligent - this was just the first part of that).

CrystalP commented:
Thanks. What does a user do to change the playlist in case the wrong one was chosen?

78andyp commented:
Choose playlist is added to the context menu.

78andyp committed:
[Video] Fix for videoversions (and other fixes) and bluray:// paths.

	modified:   xbmc/utils/SaveFileStateJob.cpp
	modified:   xbmc/utils/URIUtils.cpp
	modified:   xbmc/utils/URIUtils.h
	modified:   xbmc/video/VideoDatabase.cpp
	modified:   xbmc/video/VideoDatabase.h

78andyp committed:
Improve SetStreamDetailsForFile return value

78andyp commented:
@CrystalP - I've had a go at improving `SetStreamDetailsForFile()` returns.

CrystalP commented:
Thanks, but looking at the commit in isolation that's not what I meant. However in the context of the PR maybe it is,
I'm still wrapping my head around the changes of the initial PR and its impacts, to understand what needs to be done here.
The simplification I meant in my PR was that this fixup PR should remove the double duty of SetStreamDetailsForFile, therefore no need to convey different errors anymore.

One thing that would be very appreciated would be to separate the PR in multiple commits, one for each "logical step" in the changes. Here there are multiple fixes, one commit per fix is a good place to start. That makes review much easier. Splitting the commits is easy when each change is in a separate file, but changes in the same file must be manually split. Each commit must build successfully. During development it's best to do multiple commits, send the PR, and squash later if requested in the review. The opposite is not so easy...

CrystalP commented:
Running the code showed some issues, but first there is a big question mark on the approach.

Making the strFilename field of files an absolute path instead of a filename appended to the strPath of path is quite a big change from how things usually work.
Unexpected consequences elsewhere wouldn't be too surprising. Did you find other existing cases doing this?
Even if strFilename was relative to the path, including directories in the filename would be quite different from the norm.

> > What does a user do to change the playlist in case the wrong one was chosen?

> Choose playlist is added to the context menu (in the original PR).

That doesn't seem to be available after playing the menu though likely not created by this PR.
And after playing/resuming through the menu a couple times, the library item doesn't work anymore with "this file is no longer available. would you like to remove it".
The error is CGUIWindowVideoBase::PlayItem: File 'bluray://D%3a%5cTests%5cAlita%20Battle%20Angel%20(2019)%5c/menu' for library item 'videodb://movies/titles/1' doesn't exist.

Problem when attempting to refresh from movie info dialog - offers movies matching the name "playlist" instead of the movie name (that was after playing one of the .mpls and stopping).

I also saw some additional (and that remain unused) records created in path and files.

I haven't tried versions yet.

78andyp commented:
> Running the code showed some issues, but first there is a big question mark on the approach.
> 
> Making the strFilename field of files an absolute path instead of a filename appended to the strPath of path is quite a big change from how things usually work. Unexpected consequences elsewhere wouldn't be too surprising. Did you find other existing cases doing this? Even if strFilename was relative to the path, including directories in the filename would be quite different from the norm.
> 
> > > What does a user do to change the playlist in case the wrong one was chosen?
> 
> > Choose playlist is added to the context menu (in the original PR).
> 
> That doesn't seem to be available after playing the menu though likely not created by this PR. And after playing/resuming through the menu a couple times, the library item doesn't work anymore with "this file is no longer available. would you like to remove it". The error is CGUIWindowVideoBase::PlayItem: File 'bluray://D%3a%5cTests%5cAlita%20Battle%20Angel%20(2019)%5c/menu' for library item 'videodb://movies/titles/1' doesn't exist.
> 
> Problem when attempting to refresh from movie info dialog - offers movies matching the name "playlist" instead of the movie name (that was after playing one of the .mpls and stopping).
> 
> I also saw some additional (and that remain unused) records created in path and files.
> 
> I haven't tried versions yet.

Thanks for this.

Regarding paths. The reason I changed it here is because strPath is used to look for local art/nfo etc.. and a partial vfs path would not point in the correct location on disk.

Going back over the history of this - see #24277 - my original attempt was to create a playlist field in the database but @rmrector reminded me about the vfs and it seemed a more elegant solution to use that. There seemed to be a consensus though that the (then) current implementation didn't work well for ISOs. Should I revert to that??

Thanks for highlighting the other issues, I'll look at those if we decide this is the way forwards.

78andyp commented:
Thanks.
I don't think it should ever return 'menu' - I thought I had dealt with that. I wlil look again.
Will look at RAR/ZIP, thanks.

CrystalP commented:
After experimenting with a zip,
path strPath archive://D%3a%5cTests%5cMovies-zip-rar%5cmovies-flat.zip/Tears of Steel/
files strFilename tears_of_steel_1080p.mov

rmrector commented:
> After experimenting with a zip,
path strPath archive://D%3a%5cTests%5cMovies-zip-rar%5cmovies-flat.zip/Tears of Steel/
files strFilename tears_of_steel_1080p.mov

Yeah, that looks right.

One thing I may have left unclear, I figure the `bluray://` VFS-style URL is only necessary for ISO files. When reading a Blu-ray disc as loose files you can point directly to the playlist file as a path in the host file system, but when the disc is an ISO we need to first point to the ISO in the host file system and then point to the playlist file _inside_ the ISO.

> Regarding paths. The reason I changed it here is because strPath is used to look for local art/nfo etc.. and a partial vfs path would not point in the correct location on disk.

I see how that is still unhelpful for finding artwork and such, though. I wonder if there is an extra trick somewhere for archives to find artwork _next to the archive_, rather than inside the archive next to the movie file inside it. If not, that may be a separate effort to solve for this and archives and any other containers.

78andyp commented:
Thanks. Will rewrite this trying to address the original videoversions bug and the local directory issues but keep the filename relative to the path.

78andyp committed:
Keep file relative to path

78andyp committed:
Fix clean database deletes archive:// paths

78andyp committed:
Fix Information Dialog refresh doesn't use correct movie name

78andyp committed:
Fix Choose Playlist context menu option missing if in menu

78andyp committed:
Fix cannot play again if stopped in menu

78andyp committed:
Further menu related fixes

78andyp committed:
Fix Export bluray:// to seperate files

	modified:   xbmc/utils/ArtUtils.cpp
	modified:   xbmc/video/VideoDatabase.cpp
	modified:   xbmc/video/VideoDatabase.h

78andyp committed:
Better handling of bluray:// file and streamdetails changes

78andyp committed:
Fix episode export/import with bluray://

78andyp committed:
Fix resume when played through menu

78andyp committed:
Jenkins

78andyp commented:
@CrystalP - this should be good for review now. Thanks.

78andyp committed:
Clarify function name and small optimisation

78andyp committed:
Fix test

78andyp commented:
@KarellenX @rmrector 
Hi both, can I ask for some advice/thoughts on a couple of things I've noticed whilst stepping through some of this, and related code??

Firstly - looking at `CVideoTagLoaderNFO::FindNFO()` - this seems overly complicated and more lax than the wiki would suggest.
If passed a folder bluray (ie. <movie_name>/BDMV/index.BDMV) it first looks for BDMV/index.nfo (as per the wiki), it then recurses to look for any single NFO in the <movie_name> directory and returns to look for any single NFO in the BDMV directory - is this by design or accident?? I know according to the wiki you can have movie.nfo - but 'movie' doesn't feature in the code at all - it's any single NFO.

Secondly - I've noticed that the `.actors` folder for bluray and DVD folder structures are placed in the BDMV and VIDEO_TS directories - it seems simply because `GetParentPath()` is used v. `GetLocalMetadataPath()` - and the latter is interestingly not used for finding to NFO!! Again - is this by design or would it make more sense to have the `.actors` folder in the base movie folder with the other art?? - or would this annoy too many people with existing collections??

KarellenX commented:
Hi @78andyp 

For your first point, I'll leave that for a developer to respond.

>would it make more sense to have the .actors folder in the base movie folder with the other art?? - or would this annoy too many people with existing collections??

Yea, I think it would annoy too many people to move the .actors folder for blurays and dvd's.
Now that the actors bug for episodes has been fixed, I'd rather see a complete overhaul of the actors rather than little changes that a lot of people either wouldn't notice or would get annoyed by and additionally force 3rd party media managers to update software for no gain.

78andyp commented:
> Hi @78andyp
> 
> For your first point, I'll leave that for a developer to respond.
> 
> > would it make more sense to have the .actors folder in the base movie folder with the other art?? - or would this annoy too many people with existing collections??
> 
> Yea, I think it would annoy too many people to move the .actors folder for blurays and dvd's. Now that the actors bug for episodes has been fixed, I'd rather see a complete overhaul of the actors rather than little changes that a lot of people either wouldn't notice or would get annoyed by and additionally force 3rd party media managers to update software for no gain.

Thanks. Fair comment.

Follow on question - would you expect the art/nfo for an archive:// (eg. zip file) to be IN the archive file itself or in the directory containing the zip file?? (either way I think it's broken)

KarellenX commented:
>Follow on question - would you expect the art/nfo for an archive:// (eg. zip file) to be IN the archive file itself or in the directory containing the zip file?? (either way I think it's broken)

I've never used them.  
But I would expect it to be outside the zip file.

78andyp committed:
Fix fail to delete bluray:// file if allow delete from library enabled (along with base path function optimisation)

78andyp committed:
Rebase fixes

78andyp committed:
Further small code optimisation

garbear commented:
CC me. I'll take a look once the commit history is cleaned up and jenkins changes are addressed.

78andyp commented:
> CC me. I'll take a look once the commit history is cleaned up and jenkins changes are addressed.

Thanks @garbear. What do you mean by commit history cleaned up?? I was told it all needed to be seperated out. Do you want me to merge it all into 1?? Jenkins changes were mostly addressed in a later commit.

notspiff commented:
This is just a friendly hint: While I realize you need an 'all the changes' branch while working, it is very often good to split into many small PRs for the actual merge, even though some people will say not to do so (hope they don't read this :P). This due to a mix of psychology and practicalities. It's much easier for somebody to spend 5 mins on a review, ten times, than it is for somebody to spend 50 mins to review everything in one go. This makes everything stall, for years even and causes a lot of frustration. Let me be clear: I am NOT saying you have done anything wrong, I am not putting blame on any party, it is just how it is.

I've been flushing my (monkey) work lately and I get a commit or two in every day instead of aiming for getting the 50 commits in one go, even though they are all of the same nature (enum to enum class). Same approach could be applied here. If a commit makes things better on its own and do not cause regressions, split it, ship it and conquer.

garbear commented:
@78andyp Good question. The PR description mentions 9 fixes:

```
Fixes:
- bluray:// path not updated if more than one videoversion
- videoversions would be readded after library scan after bluray:// path assigned
- library clean coud remove bluray:// paths (and archive:// paths)
- now removes orphan `file` and `path` entries
- information diaglog refresh used wrong movie name
- choose playlist option missing if stopped in menu
- import/export of bluray:// containing items
- fix resume when played through menu
- fix could not delete bluray:// when removing from library and rename/delete enabled
```

I would expect 9 commits, 1 for each fix. Or some similar combination of commits. The reason, beyond reviewability (which is of primary importance to a reviewer), is the ability to bisect and revert problematic commits.

For example, the last fix, "Fix fail to delete bluray:// file if allow delete from library enabled (along with base path function optimisation)", cannot be reverted in case of a problem:

```
$ git revert e5cb3dbf1a6f3c4757f08cae08ef48d4c0392f03
Auto-merging xbmc/video/VideoDatabase.cpp
Auto-merging xbmc/video/dialogs/GUIDialogVideoInfo.cpp
CONFLICT (content): Merge conflict in xbmc/video/dialogs/GUIDialogVideoInfo.cpp
error: could not revert e5cb3dbf1a... Fix fail to delete bluray:// file if allow delete from library enabled (along with base path function optimisation)
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git revert --continue".
hint: You can instead skip this commit with "git revert --skip".
hint: To abort and get back to the state before "git revert",
hint: run "git revert --abort".
hint: Disable this message with "git config advice.mergeConflict false"
```

78andyp commented:
> @78andyp Good question. The PR description mentions 9 fixes:
> 
> ```
> Fixes:
> - bluray:// path not updated if more than one videoversion
> - videoversions would be readded after library scan after bluray:// path assigned
> - library clean coud remove bluray:// paths (and archive:// paths)
> - now removes orphan `file` and `path` entries
> - information diaglog refresh used wrong movie name
> - choose playlist option missing if stopped in menu
> - import/export of bluray:// containing items
> - fix resume when played through menu
> - fix could not delete bluray:// when removing from library and rename/delete enabled
> ```
> 
> I would expect 9 commits, 1 for each fix. Or some similar combination of commits. The reason, beyond reviewability (which is of primary importance to a reviewer), is the ability to bisect and revert problematic commits.
> 
> For example, the last fix, "Fix fail to delete bluray:// file if allow delete from library enabled (along with base path function optimisation)", cannot be reverted in case of a problem:
> 
> ```
> $ git revert e5cb3dbf1a6f3c4757f08cae08ef48d4c0392f03
> Auto-merging xbmc/video/VideoDatabase.cpp
> Auto-merging xbmc/video/dialogs/GUIDialogVideoInfo.cpp
> CONFLICT (content): Merge conflict in xbmc/video/dialogs/GUIDialogVideoInfo.cpp
> error: could not revert e5cb3dbf1a... Fix fail to delete bluray:// file if allow delete from library enabled (along with base path function optimisation)
> hint: After resolving the conflicts, mark them with
> hint: "git add/rm <pathspec>", then run
> hint: "git revert --continue".
> hint: You can instead skip this commit with "git revert --skip".
> hint: To abort and get back to the state before "git revert",
> hint: run "git revert --abort".
> hint: Disable this message with "git config advice.mergeConflict false"
> ```
> 
> Also, Kodi must be able to compile between every commit (and without Jenkins reporting any style issues for individual commits). Jenkins changes can't be addressed in a later commit, they must be applied to each individual commit.

Thanks.
I'm still a `git` novice - how to you alter commits within the 'commit chain'??

notspiff commented:
https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

garbear commented:
> I'm still a `git` novice - how to you alter commits within the 'commit chain'??

The best of us start from this point. You can use an "interactive rebase" to modify the commit chain, like @notspiff points out. I encourage you to struggle with git for a while, and when all else fails (as it will, while training in the art of git fu) you can do a "soft reset" on the upstream base (`git reset 4128ad358688a5daea9cb0bf52442005a6f30e1f`) which will clear all history but leave all changes. You can then add changes piecewise (`git add -p`). Worst case, you may have to redo everything by hand. And you can always undo everything by resetting to the last known good state here: `git reset --hard 8974111a94bca965c4aeb149c465eb29be310aba`.

notspiff commented:
Also; 
1) branches cost nothing, always make a backup before modifying history.
2) as long as a state has been committed, you can always recall it using 'git reflog'

78andyp commented:
Thanks both.
I think i'll start with a fresh branch.
Is there (an easy) way to apply a single commit from one branch to a new branch??

notspiff commented:
```git cherry-pick commit-id```

78andyp commented:
Thanks @notspiff 
The git manual link you kindly provided suggests do all work locally before pushing all commits at the end.
If that's the case - is there a way to run Jenkins against local commits - so as not apply one Jenkins commit at the end (as per @garbear)??

78andyp commented:
@garber - could you kindly look at #26220 and tell me if you are happy so far - I've changed some of the commits to be more logical and fit with the fixes.

CrystalP commented:
@78andyp this is next on my to-review list, thanks for your efforts and for having split the changes in more manageable chunks. Looks like more things snuck in the end though :-) I should have more time after the holidays.

CrystalP commented on xbmc/FileItem.cpp:
```diff
@@ -883,12 +883,15 @@ bool CFileItem::Exists(bool bUseCache /* = true */) const
+  {
+    strPath = CURL(strPath).GetHostName();
     return CDirectory::Exists(strPath, bUseCache);
-  else
-    return CFile::Exists(strPath, bUseCache);
+  }
```
I think the fix belongs to the bluray:// protocol handler, which should recognize the virtual "menu" file.
